### ISSUE TITLE
Confusion regarding strand and other filters

### ISSUE THREAD

**Issue by tfenne** [AstraZeneca-NGS/VarDictJava#62]
I got bitten by something recently and was curious to understand the reasoning behind some filtering choices in VarDict[Java] and whether there is any interest in re-visiting those choices.

Starting from the end, I recently discovered that VarDict discard's all variants that do not return true when evaluated with the `isGoodVar()` method.  This includes several tests that I was surprised to find, including ones for strand-bias and position-in-read.  My assumption had been that variants that are strand or position-in-read biased, but met other criteria (min reads, min fraction) would be emitted by VarDict but be filtered in the resulting VCF.  Indeed, some variants are emitted this way, but others are prematurely filtered by `isGoodVar()` and never emitted.

This came up because of a dataset that was processed two slightly different ways, in which a variant was called in one and not the other.  The variant in question is a simple bi-allelic SNP in relatively deep coverage that is clearly strand biased.  The contingency tables are as follows:

```
   processv1     processv2
   ref   alt     ref   alt
+  223     6     183     5
-  185    88     130    87
```

The p-values for fisher's exact test are `7.69e-20` and `6.98e-22` respectively; so v2 shows _more_ strand bias, but both are very significant.  In v1 the variant is emitted in the VCF and flagged for bias, whereas in v2 it is not emitted at all.

My preference would be to have VarDict emit all variants that pass the explicit criteria set on the command line (min reads, quality, min allele fraction etc.) and filter the results appropriately.  I can  do this with the current version by a) invoking pileup mode, b) filtering out invariant sites and c) filtering out sites that fail min-depth, min-maf etc., but this is rather awkward.

I'm curious what others think about this?  I'm happy to do the work to test removing some of the tests in `isGoodVar()` and ensure that the additional variants are filtered in the resulting VCF if others think that would be preferable.  Or perhaps a command line flag to disable some of these filters?

**Comment by wendymclee3** [comment]
I agree with @tfenne that it'd ideal to have the filtered variants to be emitted in the resulting VCF. I spent quite a bit of time tracking down the variants that are not called and tried to understand why VarDict wouldn't emit those calls when the min reads and min fraction criteria were met.

**Comment by zhongwulai** [comment]
@tfenne Can you double check whether the v2 was filtered due to strand bias?  Given those numbers, it should be filtered due to other parameters, not because of strand bias.  Or you can send the pipleup mode output for variant.

There're some redundancies in VarDict's "isGoodVar()" function and var2vcf_valid.pl script.  Move to vcf step might make sense.  I've been using the "-p" pileup mode for de-bugging as well.  But running in pileup mode is quite expensive.  Any further suggestions beyond (min reads, quality, min allele fraction)?

**Comment by tfenne** [comment]
@zhongwulai Thanks for the response.  I'm 99% sure that it was strand-bias filtering, because part my debugging process was to run with `-B 10000` to effectively disable the strand bias filtering and  that caused the variant to be emitted.

I've attached a file with the pileup mode results for both the v1 and v2 inputs at that locus:
[pileup.txt](https://github.com/AstraZeneca-NGS/VarDictJava/files/748193/pileup.txt).  What I do see is that the bias field is `2;2` for v1 and `2;1` for v2.

I don't fully understand the MSI related filters in `isGoodVar()`, but I think I would suggest hard filtering on:
1. Min supporting reads (`-r`)
2. Min allele fraction (`-f`)
3. Indel size (`-I`)

And would prefer to see VCF filters (soft?) applied for:
1. Min mean mapping quality (`-O`)
2. Strand bias (`-B`)
3. The Qratio test (`-o`)
4. The read position filter test (`-P`)

My reasoning (and it's just my opinion) is that the suggested hard filters are very easy for a user to comprehend, say staring at a pileup in IGV, and understand why something wasn't called, whereas the things I've suggested should always be output and filtered in the VCF require quite a bit more computation, and are hard for a user to intuit while looking at a viewer.  Being able to see the filtering reason in the filter field, and correlate that to a value in the INFO field would be quite helpful I think.

**Comment by mjafin** [comment]
@tfenne would you be able to provide a small bam file with good variants and variants falling into some of the above categories? If you also provide some meta code for the implementation we could ask our friends (@ZLyanov) to implement this, ideally without regressions.

**Comment by tfenne** [comment]
Hi @mjafin I can certainly try to put together some data that illustrates the behaviour.  I'm tempted to put together a PR as an example of what I'm talking about that because I think it might be easier than pseudo-coding.  Essentially what I'm suggesting is removing many of the filters from `isGoodVar` with the goal being:
1) More of the variants get passed through to the VCF via `var2vcf_valid.pl`
2) All the additional variants are flagged with one or more `FILTER`s
3) Leaving the set of `PASS` variants identical

**Comment by tfenne** [comment]
Thinking about this more, here's what I think I'm suggesting, as comments in the isGoodVar() function:

```java
    static boolean isGoodVar(Variant vref, Variant rref, String type,
            Set<String> splice,
            Configuration conf) {
        if (vref == null || vref.refallele == null || vref.refallele.isEmpty())
            return false;

        if (type == null || type.isEmpty()) {
            type = varType(vref.refallele, vref.varallele);
        }

        // TODO: remove the last two conditions as they are also checked in var2vcf_valid.pl
        if (vref.freq < conf.freq
                || vref.hicnt < conf.minr
                || vref.pmean < conf.readPosFilter
                || vref.qual < conf.goodq) {
            return false;
        }

        // TODO: ideally this could also be removed, but I think it is a slightly different test
        // TODO: than exists in var2vcf_valid.pl.  var2vcf could be modified to include this test
        // TODO: and generate a new filter if the variant mean mapq << ref mean mapq?
        if (rref != null && rref.hicnt > conf.minr && vref.freq < 0.25d) {
            //The reference allele has much better mean mapq than var allele, thus likely false positives
            double d = vref.mapq + vref.refallele.length() + vref.varallele.length();
            if ((d - 2 < 5 && rref.mapq > 20)
                    || (1 + d) / (rref.mapq + 1) < 0.25d) {
                return false;
            }
        }

        if (type.equals("Deletion") && splice.contains(vref.sp + "-" + vref.ep)) {
            return false;
        }
        
        // TODO: ideally move this too; qratio is available in var2vcf_valid.pl and could be filtered on there
        if (vref.qratio < conf.qratio) {
            return false;
        }
        if (vref.freq > 0.35d) {
            return true;
        }
        
        // TODO: Remove this filter as it's already replicated in var2vcf_valid.pl
        if (vref.mapq < conf.mapq) {
            return false;
        }
        if (vref.msi >= 13 && vref.freq <= 0.275d && vref.msint == 1) {
            return false;
        }
        if (vref.msi >= 8 && vref.freq <= 0.2d && vref.msint > 1) {
            return false;
        }
        
        // TODO: Remove this filter as it's already replicated in var2vcf_valid.pl (I think)
        if (vref.bias.equals("2;1") && vref.freq < 0.25d) {
            if (type == null || type.equals("SNV") || (vref.refallele.length() <= 3 && vref.varallele.length() <= 3)) {
                return false;
            }
        }

        return true;
    }
```
